gwcs/wcs.py

```diff
@@ -794,6 +794,132 @@ def to_fits_sip(self, bounding_box=None, max_pix_error=0.25, degree=None,
+        -------
+        hdr : astropy.io.fits.Header
+            Header with WCS-TAB information associated (to be used) with image
+            data.
+
+        bin_table : astropy.io.fits.BinTableHDU
+            Binary table extension containing the coordinate array.
+
+        Raises
+        ------
+        ValueError
+            When ``bounding_box`` is not defined either through the input
+            ``bounding_box`` parameter or this object's ``bounding_box``
+            property.
+
+        ValueError
+            When the number of ``sampling`` values is not 1 and also
+            does not match the number of image axes.
+
+        RuntimeError
+            If the number of image axes (`~gwcs.WCS.pixel_n_dim`) is larger
+            than the number of world axes (`~gwcs.WCS.world_n_dim`).
+
+        """
+        if bounding_box is None:
+            if self.bounding_box is None:
+                raise ValueError(
+                    "Need a valid bounding_box to compute the footprint."
+                )
+            bounding_box = self.bounding_box
+
+        if self.pixel_n_dim > self.world_n_dim:
+            raise RuntimeError(
+                "The case when the number of input axes is larger than the "
+                "number of output axes is not supported."
+            )
+
+        # 1D grid coordinates:
+        gcrds = []
+        cdel = []
+        sampling = np.atleast_1d(sampling).ravel()
+        if sampling.size != self.pixel_n_dim:
+            if sampling.size == 1:
+                sampling = np.repeat(sampling, self.pixel_n_dim)
+            else:
+                raise ValueError("Number of sampling values either must be 1 "
+                                 "or it must match the number of pixel axes.")
+
+        for (xmin, xmax), s in zip(bounding_box, sampling):
+            npix = max(2, 1 + int(np.ceil(abs((xmax - xmin) / s))))
+            gcrds.append(np.linspace(xmin, xmax, npix))
+            cdel.append((xmax - xmin) / (npix - 1))
+
+        # n-dim coordinate arrays:
+        coord = np.stack(
+            self(*np.meshgrid(*gcrds[::-1], indexing='ij')[::-1]),
+            axis=-1
+        )
+
+        # create primary header:
+        hdr = fits.Header()
+
+        for k in range(self.world_n_dim):
+            k1 = k + 1
+            ct = cf.get_ctype_from_ucd(self.world_axis_physical_types[k])
+            if len(ct) > 4:
+                raise ValueError("Axis type name too long.")
+
+            hdr['CTYPE{:d}'.format(k1)] = ct + (4 - len(ct)) * '-' + '-TAB'
+            hdr['CUNIT{:d}'.format(k1)] = self.world_axis_units[k]
+            hdr['PS{:d}_0'.format(k1)] = bin_ext_name
+            hdr['PS{:d}_1'.format(k1)] = crd_col_name
+            hdr['PV{:d}_3'.format(k1)] = k1
+            hdr['CRVAL{:d}'.format(k1)] = 1
+
+            if k < self.pixel_n_dim:
+                hdr['CRPIX{:d}'.format(k1)] = gcrds[k][0] + 1
+                hdr['CD{0:d}_{0:d}'.format(k1)] = 1.0 / cdel[k]
```

Comment:
Is this creating a CD matrix? We should use PC instead as it is the standard for FITS. It may be the identity matrix and then CDELTs will have values != 1.

Reply:
From Fig. 6 in paper III (the one with `-TAB`) I did not get the feeling that `CD` is discouraged, but if you prefer `PC` - I am fine with it.